File: testdata/precision-suite-clojure/tp_codeinjection.clj
(ns app.codeinjection
  "Code injection (CWE-95): an untrusted value flows into eval / load-string,
   which re-evaluate their argument as Clojure code. A correct scanner fires
   TAINT-005.")

;; eval of a request-derived form.
(defn run-form [req]
  (let [form (:params req)]
    (eval form))) ; nox-expect: TAINT-005

;; load-string reads and evaluates a whole program string from the query string.
(defn run-script [req]
  (let [s (:query-string req)]
    (load-string s))) ; nox-expect: TAINT-005

File: testdata/precision-suite-elixir/tp_codeinjection.exs
# Code injection (CWE-95): an untrusted value is evaluated as Elixir source via
# Code.eval_string, and untrusted bytes are deserialized via
# :erlang.binary_to_term (CWE-502). A correct scanner fires TAINT-005 on each.

defmodule TpCodeInjection do
  # Code.eval_string evaluates its argument as Elixir source.
  def evaluate(conn) do
    code = conn.params["expr"]
    Code.eval_string(code) # nox-expect: TAINT-005
  end

  # :erlang.binary_to_term deserializes untrusted bytes into terms.
  def deserialize(conn) do
    blob = conn.body_params
    :erlang.binary_to_term(blob) # nox-expect: TAINT-005
  end
end
